watchdog.go

```diff
@@ -26,239 +76,316 @@ var (
-			ch := gcTriggered
-			gcTriggered = nil
-			lk.Unlock()
+	policy, err := policyCtor(limit)
+	if err != nil {
+		return fmt.Errorf("failed to construct policy with limit %d: %w", limit, err), nil
+	}
 
-			// close and drain
-			close(ch)
-			for range ch {
-			}
-			return
+	_watchdog.scope = UtilizationSystem
+	_watchdog.closing = make(chan struct{})
+
+	var gcTriggered = make(chan struct{}, 16)
+	setupGCSentinel(gcTriggered)
+
+	_watchdog.wg.Add(1)
+	go func() {
+		defer _watchdog.wg.Done()
+
+		var (
+			memstats  runtime.MemStats
+			sysmem    gosigar.Mem
+			threshold uint64
+		)
+
+		// initialize the threshold.
+		threshold, immediate := policy.Evaluate(UtilizationSystem, sysmem.ActualUsed)
+		if immediate {
+			Logger.Warnf("system-driven watchdog requested immediate GC upon startup; " +
+				"policy is probably misconfigured; " +
+				"performance compromised")
+			forceGC(&memstats)
 		}
 
-		// ReadMemStats stops the world. But as of go1.9, it should only
-		// take ~25µs to complete.
-		//
-		// Before go1.15, calls to ReadMemStats during an ongoing GC would
-		// block due to the worldsema lock. As of go1.15, this was optimized
-		// and the runtime holds on to worldsema less during GC (only during
-		// sweep termination and mark termination).
-		//
-		// For users using go1.14 and earlier, if this call happens during
-		// GC, it will just block for longer until serviced, but it will not
-		// take longer in itself. No harm done.
-		//
-		// Actual benchmarks
-		// -----------------
-		//
-		// In Go 1.15.5, ReadMem with no ongoing GC takes ~27µs in a MBP 16
-		// i9 busy with another million things. During GC, it takes an
-		// average of less than 175µs per op.
-		//
-		// goos: darwin
-		// goarch: amd64
-		// pkg: github.com/filecoin-project/lotus/api
-		// BenchmarkReadMemStats-16                	   44530	     27523 ns/op
-		// BenchmarkReadMemStats-16                	   43743	     26879 ns/op
-		// BenchmarkReadMemStats-16                	   45627	     26791 ns/op
-		// BenchmarkReadMemStats-16                	   44538	     26219 ns/op
-		// BenchmarkReadMemStats-16                	   44958	     26757 ns/op
-		// BenchmarkReadMemStatsWithGCContention-16    	      10	    183733 p50-ns	    211859 p90-ns	    211859 p99-ns
-		// BenchmarkReadMemStatsWithGCContention-16    	       7	    198765 p50-ns	    314873 p90-ns	    314873 p99-ns
-		// BenchmarkReadMemStatsWithGCContention-16    	      10	    195151 p50-ns	    311408 p90-ns	    311408 p99-ns
-		// BenchmarkReadMemStatsWithGCContention-16    	      10	    217279 p50-ns	    295308 p90-ns	    295308 p99-ns
-		// BenchmarkReadMemStatsWithGCContention-16    	      10	    167054 p50-ns	    327072 p90-ns	    327072 p99-ns
-		// PASS
-		//
-		// See: https://github.com/golang/go/issues/19812
-		// See: https://github.com/prometheus/client_golang/issues/403
-
-		if eventIsGc {
-			config.Logger.Infof("watchdog after GC")
+		for {
+			select {
+			case <-Clock.After(frequency):
```

Comment:
Can we avoid creating new timers every time? Also, this won't _cancel_ the timer if we're frequently firing GC (we could leak timers).

Reply:
Yes, requires some more work to reset the timer, and the associated dance, but can be done.

Reply:
Fixed.